File: src/routes/replit.ts
import { type Request, type Response, NextFunction } from "express";
import { Router } from "express";
import { authMiddleware } from "../middleware/auth";
import { pool } from "../config/pg";
import mongoose from "mongoose";
import { v4 as uuidv4 } from "uuid";
import dotenv from "dotenv";
import { Project, IProject } from "../models/Project";
import fs from "fs";
import path from "path";
import { exec } from "child_process";
import { generateToken } from "../utils/jwt";
import bcrypt from "bcrypt";

dotenv.config();
const router = Router();

const WORKSPACE_DIR = path.join(
  process.cwd(),
  process.env.WORKSPACE_DIR || "workspaces"
);

if (!fs.existsSync(WORKSPACE_DIR)) {
  fs.mkdirSync(WORKSPACE_DIR);
}
// ===========Authenticated Routes==========

router.post("/register", async (req: Request, res: Response) => {
  const { email, password } = req.body;
  if (!email || !password) {
    res.status(400).json({ message: "Email and password are required" });
    return;
  }
  const hashedPassword = await bcrypt.hash(password, 10);
  const id = uuidv4();
  await pool.query(
    "INSERT INTO users (id, email, password) VALUES ($1,$2,$3)",
    [id, email, hashedPassword]
  );

  res.json({ message: "User registered" });
});

router.post("/login", async (req: Request, res: Response) => {
  const { email, password } = req.body;
  if (!email || !password) {
    res.status(400).json({ message: "Email and password are required" });
    return;
  }
  const checkUser = await pool.query("SELECT * FROM users WHERE email = $1", [
    email,
  ]);
  if (checkUser.rowCount === 0) {
    res.status(400).json({ message: "Invalid credentials" });
    return;
  }
  const user = checkUser.rows[0];
  const isMatch = await bcrypt.compare(password, user.password);
  if (!isMatch) {
    res.status(400).json({ message: "Invalid credentials" });
    return;
  }
  const token = generateToken({ userId: user.id, email: user.email });
  res.json({ token });
});

/* ================= PROJECT ================= */

router.post(
  "/create-project",
  authMiddleware,
  async (req: Request, res: Response) => {
    const userId = (req as any).user.userId;
    const { dependencies = {} } = req.body;

    const existing = await Project.findOne({ userId });
    if (existing) {
      res.status(400).json({ message: "Project already exists" });
      return;
    }

    const projectId = uuidv4();
    const folderName = `${userId}_${projectId}`;
    const projectPath = path.join(WORKSPACE_DIR, folderName);
    fs.mkdirSync(projectPath, { recursive: true });
    fs.writeFileSync(
      path.join(projectPath, "package.json"),
      JSON.stringify(
        {
          name: folderName,
          version: "1.0.0",
          type: "module",
          main: "index.js",
          dependencies,
        },
        null,
        2
      )
    );

    fs.writeFileSync(
      path.join(projectPath, "index.js"),
      `
import express from "express";
const app = express();
app.get("/", (req,res)=>res.json({message:"Hello"}));
app.listen(3000);
`
    );

    const containerName = `repl-${userId}`;
    const newProject = new Project({
      userId,
      projectId,
      containerName,
    });
    await newProject.save();

    res.json({ projectId });
  }
);

//iteration1
// router.post("/run", authMiddleware, async (req: Request, res: Response) => {
//   const userId = (req as any).user.userId;
//   const project = await Project.findOne({ userId });
//   if (!project) {
//     res.status(404).json({ message: "Project not found" });
//     return;
//   }

//   const folderName = `${userId}_${project.projectId}`;
//   const projectPath = path.join(WORKSPACE_DIR, folderName);
//   const command = `
//     docker run -d \
//     --name ${project.containerName} \
//     -v ${projectPath}:/workspace \
//     -p 3000 \
//     --memory=256m \
//     --cpus=0.5 \
//     mini-replit-node \
//     nix develop --command sh -c "if [ ! -d node_modules ]; then npm install; fi && node index.js"
//   `;

//   exec(command, async (error) => {
//     if (error) {
//       console.error("Error running container:", error.message);
//       res.status(500).json({ message: "Error starting container" });
//       return;
//     }

//     setTimeout(() => {
//       //   const inspectCommand = `docker inspect ${project.containerName} --format='{{(index (index .NetworkSettings.Ports "3000/tcp") 0).HostPort}}'`;

//       const inspectCommand = `docker inspect ${project.containerName} --format="{{if (index .NetworkSettings.Ports \\"3000/tcp\\")}}{{(index (index .NetworkSettings.Ports \\"3000/tcp\\") 0).HostPort}}{{end}}"`;
//       exec(inspectCommand, async (err, stdout) => {
//         if (err) {
//           console.error("Error fetching port:", err.message);
//           res
//             .status(500)
//             .json({ message: "Container started but failed to fetch port" });
//           return;
//         }

//         const mappedPort = stdout.trim();

//         // Save port in MongoDB
//         project.port = Number(mappedPort);
//         await project.save();

//         res.json({
//           message: "Container started",
//           port: mappedPort,
//         });
//       });
//     }, 1000); // small delay
//   });
// });

//iteraton2 - added retry logic for port fetching
// router.post("/run", authMiddleware, async (req: Request, res: Response) => {
//   const userId = (req as any).user.userId;
//   const project = await Project.findOne({ userId });

//   if (!project) {
//     res.status(404).json({ message: "Project not found" });
//     return;
//   }

//   const folderName = `${userId}_${project.projectId}`;
//   const projectPath = path.join(WORKSPACE_DIR, folderName);

//   const command = `
//     docker run -d \
//     --name ${project.containerName} \
//     -v ${projectPath}:/workspace \
//     -p 3000:3000 \
//     --memory=256m \
//     --cpus=0.5 \
//     mini-replit-node \
//     nix develop --command sh -c "if [ ! -d node_modules ]; then npm install; fi && node index.js"
//   `;

//   exec(command, async (error) => {
//     if (error) {
//       console.error("Error running container:", error.message);
//       res.status(500).json({ message: "Error starting container" });
//       return;
//     }

//     // Wait a bit to ensure container is up
//     setTimeout(() => {
//       const inspectCommand = `docker inspect ${project.containerName} --format="{{if (index .NetworkSettings.Ports \\"3000/tcp\\")}}{{(index (index .NetworkSettings.Ports \\"3000/tcp\\") 0).HostPort}}{{end}}"`;

//       exec(inspectCommand, async (err, stdout) => {
//         if (err) {
//           console.error("Error fetching port:", err.message);
//           res.status(500).json({
//             message: "Container started but failed to fetch port",
//           });
//           return;
//         }

//         const mappedPort = stdout.trim();
//         if (!mappedPort) {
//           res.status(500).json({ message: "Port not yet assigned" });
//           return;
//         }

//         // Save port in MongoDB
//         project.port = Number(mappedPort);
//         await project.save();

//         res.json({
//           message: "Container started",
//           port: mappedPort,
//         });
//       });
//     }, 3000); // 3 seconds
//   });
// });

router.post("/run", authMiddleware, async (req: Request, res: Response) => {
  const userId = (req as any).user.userId;
  const project = await Project.findOne({ userId });

  if (!project) {
    return res.status(404).json({ message: "Project not found" });
  }

  const folderName = `${userId}_${project.projectId}`;
  const projectPath = path.join(WORKSPACE_DIR, folderName);

  // --- Automatically create flake.nix if it doesn't exist ---
  const flakePath = path.join(projectPath, "flake.nix");
  if (!fs.existsSync(flakePath)) {
    const flakeContent = `{
  description = "Node.js Dev Environment";

  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.11";
  };

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";
      pkgs = import nixpkgs { inherit system; };
    in
    {
      devShells.\${system}.default = pkgs.mkShell {
        buildInputs = [
          pkgs.nodejs_18
          pkgs.nodePackages.npm
        ];
      };
    };
}`;
    fs.writeFileSync(flakePath, flakeContent, "utf-8");
    console.log("flake.nix created at", flakePath);
  }

  console.log(project.containerName);
  console.log(projectPath);

  // --- Docker command ---
  //   const command = `
  //     docker run -d \
  //     --name ${project.containerName} \
  //     -v ${projectPath}:/workspace \
  //     -p 3000:3000 \
  //     --memory=256m \
  //     --cpus=0.5 \
  //     mini-replit-node
  //   `;->this wont work in windows as the path format is different, so we will use a more compatible command without volume for now and we can add volume later with some tweaks

  const command = `docker run -d --name ${project.containerName} -v "${projectPath}:/workspace" -p 3000:3000 --memory=256m --cpus=0.5 mini-replit-node`;

  exec(command, async (error) => {
    if (error) {
      console.error("Error starting container:", error.message);
      return res.status(500).json({ message: "Failed to start container" });
    }

    // Wait a few seconds to ensure container is running
    setTimeout(() => {
      exec(
        `docker ps -f name=${project.containerName}`,
        async (psErr, stdout) => {
          if (psErr || !stdout.includes(project.containerName)) {
            console.error("Container failed to start or exited immediately");
            return res
              .status(500)
              .json({ message: "Container failed to start" });
          }

          // Container is running → fetch port (already mapped 3000:3000)
          project.port = 3000;
          await project.save();

          res.json({
            message: "Container started successfully",
            port: project.port,
          });
        }
      );
    }, 3000); // give Docker a few seconds
  });
});
router.post("/stop", authMiddleware, async (req: Request, res: Response) => {
  const userId = (req as any).user.userId;
  const project = await Project.findOne({ userId });

  if (!project) {
    res.status(404).json({ message: "Project not found" });
    return;
  }

  const stopCommand = `docker stop ${project.containerName} && docker rm ${project.containerName}`;

  exec(stopCommand, async (error) => {
    if (error) {
      console.error("Error stopping container:", error.message);
      res.status(500).json({ message: "Failed to stop container" });
      return;
    }

    // Clear saved port
    project.port = undefined;
    await project.save();

    res.json({ message: "Container stopped and removed" });
  });
});

export default router;

File: python_version/app/routes/replit.py
import asyncio
import json
import os
import subprocess
import uuid
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_db, settings
from app.middleware.auth import get_current_user
from app.models import Project, User
from app.schemas.auth import LoginRequest, RegisterRequest
from app.schemas.project import CreateProjectRequest, CreateProjectResponse
from app.utils.jwt import generate_token
from app.utils.security import hash_password, verify_password

router = APIRouter()

WORKSPACE_DIR = Path(settings.workspace_dir).resolve()
if not WORKSPACE_DIR.is_absolute():
    WORKSPACE_DIR = Path.cwd() / settings.workspace_dir
WORKSPACE_DIR.mkdir(parents=True, exist_ok=True)


# ---------- Auth (no JWT required) ----------


@router.post("/register")
async def register(
    body: RegisterRequest,
    db: AsyncSession = Depends(get_db),
):
    hashed = hash_password(body.password)
    user = User(email=body.email, password=hashed)
    db.add(user)
    try:
        await db.flush()
        await db.refresh(user)
    except IntegrityError:
        raise HTTPException(status_code=400, detail="Email already registered")
    return {"message": "User registered"}


@router.post("/login")
async def login(
    body: LoginRequest,
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(User).where(User.email == body.email))
    user = result.scalar_one_or_none()
    if not user:
        raise HTTPException(status_code=400, detail="Invalid credentials")
    if not verify_password(body.password, user.password):
        raise HTTPException(status_code=400, detail="Invalid credentials")
    token = generate_token({"userId": str(user.id), "email": user.email})
    return {"token": token}


# ---------- Project (JWT required) ----------


@router.post("/create-project", response_model=CreateProjectResponse)
async def create_project(
    body: CreateProjectRequest,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    user_id = current_user["userId"]
    result = await db.execute(select(Project).where(Project.user_id == uuid.UUID(user_id)))
    existing = result.scalar_one_or_none()
    if existing:
        raise HTTPException(status_code=400, detail="Project already exists")

    project_id = str(uuid.uuid4())
    folder_name = f"{user_id}_{project_id}"
    project_path = WORKSPACE_DIR / folder_name
    project_path.mkdir(parents=True, exist_ok=True)

    dependencies = body.dependencies or {}
    package_json = {
        "name": folder_name,
        "version": "1.0.0",
        "type": "module",
        "main": "index.js",
        "dependencies": dependencies,
    }
    (project_path / "package.json").write_text(
        json.dumps(package_json, indent=2), encoding="utf-8"
    )

    index_js = """
import express from "express";
const app = express();
app.get("/", (req,res)=>res.json({message:"Hello"}));
app.listen(3000);
"""
    (project_path / "index.js").write_text(index_js.strip(), encoding="utf-8")

    container_name = f"repl-{user_id}"
    project = Project(
        user_id=uuid.UUID(user_id),
        project_id=project_id,
        container_name=container_name,
    )
    db.add(project)
    await db.flush()
    return CreateProjectResponse(projectId=project_id)


@router.post("/run")
async def run_project(
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    user_id = current_user["userId"]
    result = await db.execute(select(Project).where(Project.user_id == uuid.UUID(user_id)))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    folder_name = f"{user_id}_{project.project_id}"
    project_path = WORKSPACE_DIR / folder_name

    flake_path = project_path / "flake.nix"
    if not flake_path.exists():
        flake_content = """{
  description = "Node.js Dev Environment";

  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.11";
  };

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";
      pkgs = import nixpkgs { inherit system; };
    in
    {
      devShells.${system}.default = pkgs.mkShell {
        buildInputs = [
          pkgs.nodejs_18
          pkgs.nodePackages.npm
        ];
      };
    };
}"""
        flake_path.write_text(flake_content, encoding="utf-8")

    project_path_str = str(project_path)
    if os.name == "nt":
        project_path_str = project_path_str.replace("\\", "/")
    command = [
        "docker",
        "run",
        "-d",
        "--name", project.container_name,
        "-v", f"{project_path_str}:/workspace",
        "-p", "3000:3000",
        "--memory=256m",
        "--cpus=0.5",
        "mini-replit-node",
    ]

    try:
        subprocess.run(command, check=True, capture_output=True, text=True)
    except subprocess.CalledProcessError as e:
        raise HTTPException(
            status_code=500,
            detail="Failed to start container",
        )

    await asyncio.sleep(3)

    check = subprocess.run(
        ["docker", "ps", "-f", f"name={project.container_name}"],
        capture_output=True,
        text=True,
    )
    if check.returncode != 0 or project.container_name not in (check.stdout or ""):
        raise HTTPException(
            status_code=500,
            detail="Container failed to start",
        )

    project.port = 3000
    await db.flush()
    return {
        "message": "Container started successfully",
        "port": project.port,
    }


@router.post("/stop")
async def stop_project(
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    user_id = current_user["userId"]
    result = await db.execute(select(Project).where(Project.user_id == uuid.UUID(user_id)))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    stop_cmd = f"docker stop {project.container_name} && docker rm {project.container_name}"
    try:
        subprocess.run(stop_cmd, shell=True, check=True, capture_output=True, text=True)
    except subprocess.CalledProcessError:
        raise HTTPException(status_code=500, detail="Failed to stop container")

    project.port = None
    await db.flush()
    return {"message": "Container stopped and removed"}
